### PR TITLE
add nydus middleware

### PR DESCRIPTION
**Pull request by MinerYang**
Signed-off-by: yminer <yminer@vmware.com>

remove comments

Signed-off-by: yminer <yminer@vmware.com>

update ut manifest

Signed-off-by: yminer <yminer@vmware.com>

Thank you for contributing to Harbor!

# Comprehensive Summary of your change

# Issue being fixed
https://github.com/goharbor/harbor/issues/16951

Please indicate you've done the following:
- [ ] Well Written Title and Summary of the PR
- [ ] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [ ] Accepted the DCO. Commits without the DCO will delay acceptance.
- [ ] Made sure tests are passing and test coverage is added if needed.
- [ ] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

File: src/server/middleware/nydus/nydus.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+		if err != nil {
+			return err
+		}
+		mf := &v1.Manifest{}
+		if err := json.Unmarshal(payload, mf); err != nil {
+			return err
+		}
+
+		if isNydus {
+			subjectArt, err := artifact.Ctl.GetByReference(ctx, info.Repository, mf.Annotations["io.goharbor.artifact.v1alpha1.acceleration.source.digest"], nil)
```
Comment (chlins) [inline src/server/middleware/nydus/nydus.go:112]: The result from `distribution.UnmarshalManifest` can not get this `io.goharbor.artifact.v1alpha1.acceleration.source.digest` annotation?

Comment (MinerYang) [reply src/server/middleware/nydus/nydus.go:112]: `io.goharbor.artifact.v1alpha1.acceleration.source.digest` mark the subject/source artifact digest and it returns from nydus

Comment (MinerYang) [reply src/server/middleware/nydus/nydus.go:112]: `distribution.UnmarshalManifest` manifest reference contains the descriptor of each layers, but the annotation we need to got is from the payload., Thus we got the `manifest.Payload()`, then Unmarshal it.

File: src/server/middleware/nydus/nydus.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+		manifest, desc, err := distribution.UnmarshalManifest(contentType, body)
+		if err != nil {
+			logger.Errorf("unmarshal manifest failed, error: %v", err)
+			return err
+		}
+
+		var isNydus bool
+		for _, descriptor := range manifest.References() {
+			annotationMap := descriptor.Annotations
+			if _, ok := annotationMap["containerd.io/snapshot/nydus-bootstrap"]; ok {
```
Comment (wy65701436) [inline src/server/middleware/nydus/nydus.go:93]: define an variable for `containerd.io/snapshot/nydus-bootstrap`.

**Review by wy65701436 — APPROVED** [review]
lgtm

File: src/server/middleware/nydus/nydus.go (hunk 2 above)
Comment (MinerYang) [reply src/server/middleware/nydus/nydus.go]: updated

**Pull request merged**